src/string/strlen.rs

```diff
@@ -3,12 +3,34 @@ use core::usize;
 use c_types::*;
 
 #[no_mangle]
-pub unsafe extern "C" fn strlen(s: *const c_schar) -> size_t {
-    // TODO(adam) convert to checking word-size chunks
-    for i in 0.. {
-        if *s.offset(i) == 0 {
-            return i as usize;
+pub unsafe extern "C" fn strlen(mut s: *const c_schar) -> size_t {
+    macro_rules! align { () => { core::mem::size_of::<size_t>() } }
+    macro_rules! ones { () => { !0usize } }
+    macro_rules! highs { () => { usize::wrapping_mul(ones!(), usize::MAX)/2 + 1 } }
+    macro_rules! has_zero { ($x:expr) => {{
+        let x: usize = $x;
+        let ones: usize = ones!();
+        let highs: usize = highs!();
+        usize::wrapping_sub(x, ones & !x & highs) != 0
+    }} }
+    let a = s;
+    let mut w: *const size_t;
+    // check zero until align boundary
+    while s as usize % align!() != 0 {
+        if *s == 0 {
+            return s as usize - a as usize
         }
+        s = s.offset(1)
     }
-    return usize::MAX;
+    // check with word-chunks
+    w = s as _;
```

Comment:
Not sure why GitHub flagged my comment here as outdated -- I don't think that inferred casts are a good idea when it might make it difficult to track what conversions are happening.